File: app/src/main/java/com/msaggik/algorithmsdatastructures/java/graph/graph_tree/BreadthFirstSearch.java
package com.msaggik.algorithmsdatastructures.java.graph.graph_tree;

import com.msaggik.algorithmsdatastructures.java.graph.graph_tree.data_structures.Queue;
import com.msaggik.algorithmsdatastructures.java.graph.graph_tree.data_structures.QueueImpl;
import com.msaggik.algorithmsdatastructures.java.graph.graph_tree.entity.GraphTree;

/**
 * Алгоритм обхода дерева графа в ширину
 * Упрощённая модель:  V - количество вершин, E - количество рёбер
 * применим к задачам:
 * - проверить наличия пути из А в Б
 * - является ли неориентированный граф связным
 * - найти кратчайший путь до узла удовлетворяющего условию (в невзвешенном графе)
 *
 * Вычислительная сложность: О(V + E)
 **/

public class BreadthFirstSearch {
    public static int iterativelySumGraph(GraphTree tree) {
        Queue<GraphTree> queue = new QueueImpl<>();
        queue.offer(tree);

        int sumGraph = 0;

        while (!queue.isEmpty()) {
            GraphTree treeBuf = queue.poll();

            assert treeBuf != null;
            sumGraph += treeBuf.getVertex();

            if(treeBuf.getRight() != null) {
                queue.offer(treeBuf.getRight());
            }

            if(treeBuf.getLeft() != null) {
                queue.offer(treeBuf.getLeft());
            }
        }
        return sumGraph;
    }
}

File: app/src/main/java/com/msaggik/algorithmsdatastructures/kotlin/graph/graph_tree/BreadthFirstSearch.kt
package com.msaggik.algorithmsdatastructures.kotlin.graph.graph_tree

import com.msaggik.algorithmsdatastructures.kotlin.graph.graph_tree.data_structures.Queue
import com.msaggik.algorithmsdatastructures.kotlin.graph.graph_tree.data_structures.QueueImpl
import com.msaggik.algorithmsdatastructures.kotlin.graph.graph_tree.entity.GraphTree

object BreadthFirstSearch {
    fun iterativelySumGraph(tree: GraphTree): Int {
        val queue: Queue<GraphTree> = QueueImpl()
        queue.offer(tree)

        var sumGraph = 0

        while (!queue.isEmpty) {
            val treeBuf = checkNotNull(queue.poll())

            sumGraph += treeBuf.vertex

            if (treeBuf.right != null) {
                queue.offer(treeBuf.right)
            }

            if (treeBuf.left != null) {
                queue.offer(treeBuf.left)
            }
        }
        return sumGraph
    }
}